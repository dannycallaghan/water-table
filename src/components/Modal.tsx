import { ReactNode, memo, useEffect, useRef } from 'react';
import { createPortal } from 'react-dom';

const Modal = memo(function Modal({ children }: { children: ReactNode }) {
  const elRef = useRef<null | HTMLElement>(null);

  if (!elRef.current) {
    const div = document.createElement('div');
    elRef.current = div;
  }

  useEffect(() => {
    const modalRoot = document.getElementById('modal');
    modalRoot?.appendChild(elRef.current as HTMLElement);

    return () => {
      function remove() {
        modalRoot?.removeChild(elRef.current as HTMLElement);
      }
      remove();
    };
  }, []);

  return createPortal(<div>{children}</div>, elRef.current);
});

export default Modal;
