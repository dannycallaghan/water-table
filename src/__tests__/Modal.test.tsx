import { expect, test, describe } from 'vitest';
import { render } from '@testing-library/react';
import Modal from '../components/Modal.tsx';

describe('Modal', () => {
  test('renders correctly', () => {
    const { asFragment } = render(
      <Modal>
        <h1>Hello World</h1>
      </Modal>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
