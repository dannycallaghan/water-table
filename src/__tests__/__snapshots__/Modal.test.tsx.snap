// Vitest Snapshot v1

exports[`Modal > renders correctly 1`] = `<DocumentFragment />`;
